File: sdks/python/src/opik/integrations/otel/distributed_trace.py
import logging
from typing import Dict, Optional, TYPE_CHECKING

from opik import id_helpers
from opik.integrations.otel import types as otel_types
from opik.integrations.otel import attributes as otel_attributes

if TYPE_CHECKING:
    from opentelemetry import trace

LOGGER = logging.getLogger(__name__)

OPIK_TRACE_ID_HEADER = "opik_trace_id"
OPIK_PARENT_SPAN_ID_HEADER = "opik_parent_span_id"


def _get_header(http_headers: Dict[str, str], name: str) -> Optional[str]:
    target = name.lower()
    for key, value in http_headers.items():
        if isinstance(key, str) and key.lower() == target:
            return value
    return None


def extract_opik_distributed_trace_attributes(
    http_headers: Dict[str, str],
) -> Optional[otel_types.OpikDistributedTraceAttributes]:
    """
    Extracts Opik distributed trace attributes from HTTP headers.

    This function retrieves distributed tracing attributes from the provided
    HTTP headers using the keys `opik_trace_id` and optionally
    `opik_parent_span_id` (case-insensitive). Values are trimmed; if
    `opik_trace_id` is missing, blank, or not a valid UUID the function
    returns None (and logs a warning when invalid, or when
    `opik_parent_span_id` was provided without a `opik_trace_id`). A blank
    or non-UUID `opik_parent_span_id` is dropped with a warning while the
    valid `opik_trace_id` is still returned.

    Args:
        http_headers: A dictionary containing HTTP headers as
            keys and their corresponding values.

    Returns:
        An object containing the extracted `opik_trace_id` and `opik_parent_span_id` if available.
        Returns None if `opik_trace_id` is not found or is not a valid UUID.
    """
    raw_trace_id = _get_header(http_headers, OPIK_TRACE_ID_HEADER)
    trace_id = raw_trace_id.strip() if isinstance(raw_trace_id, str) else None

    raw_parent_span_id = _get_header(http_headers, OPIK_PARENT_SPAN_ID_HEADER)
    parent_span_id = (
        raw_parent_span_id.strip() if isinstance(raw_parent_span_id, str) else None
    )

    if not trace_id:
        if parent_span_id:
            LOGGER.warning(
                "Opik distributed trace header '%s' is missing while '%s' "
                "is provided; skipping distributed trace processing.",
                OPIK_TRACE_ID_HEADER,
                OPIK_PARENT_SPAN_ID_HEADER,
            )
        return None
    if not id_helpers.is_valid_uuid_v7(trace_id):
        LOGGER.warning(
            "Opik distributed trace header '%s' is not a valid UUIDv7; "
            "skipping distributed trace processing.",
            OPIK_TRACE_ID_HEADER,
        )
        return None

    if not parent_span_id:
        parent_span_id = None
    elif not id_helpers.is_valid_uuid_v7(parent_span_id):
        LOGGER.warning(
            "Opik distributed trace header '%s' is not a valid UUIDv7; "
            "ignoring parent span id.",
            OPIK_PARENT_SPAN_ID_HEADER,
        )
        parent_span_id = None

    return otel_types.OpikDistributedTraceAttributes(
        opik_trace_id=trace_id,
        opik_parent_span_id=parent_span_id,
    )


def attach_to_parent(span: "trace.Span", http_headers: Dict[str, str]) -> bool:
    """
    Attaches a distributed trace parent to the provided span, using extracted
    headers to gather the necessary trace-related attributes.

    This function parses the provided HTTP headers to extract OPIK distributed trace
    headers. If valid headers are found, it converts them into OpenTelemetry span
    attributes and attaches them to the span. If no valid headers are found, the
    function returns False.

    Args:
        span: A `opentelemetry.trace.Span` object to which the distributed trace parent will be
            attached.
        http_headers: A dictionary of HTTP headers where distributed trace headers
            may be present.

    Returns:
        True if the distributed trace headers were successfully extracted
            and attached, otherwise False.
    """
    opik_distributed_trace_headers = extract_opik_distributed_trace_attributes(
        http_headers
    )
    if opik_distributed_trace_headers is None:
        return False

    attributes = opik_distributed_trace_headers.as_attributes()
    # Mint a stable opik.span_id for this boundary span so descendants picked up by
    # OpikSpanProcessor can chain through it (their opik.parent_span_id will reference
    # this value, and the backend uses opik.span_id verbatim — see OpenTelemetryMapper).
    attributes[otel_attributes.OPIK_SPAN_ID] = id_helpers.generate_id()
    span.set_attributes(attributes)
    return True

File: sdks/typescript/src/opik/integrations/opik-otel/src/distributedTrace.ts
import { generateId, logger } from "opik";
import { OPIK_SPAN_ID } from "./attributes";
import {
  OPIK_PARENT_SPAN_ID_HEADER,
  OPIK_TRACE_ID_HEADER,
  isValidUuidV7,
} from "./internal";
import { OpikDistributedTraceAttributes } from "./OpikDistributedTraceAttributes";

/**
 * Minimal structural type for an OpenTelemetry `Span`. Only the
 * `setAttributes` method is required for distributed-trace propagation, so we
 * avoid taking a runtime dependency on `@opentelemetry/api`. Any object that
 * implements this method — including `import("@opentelemetry/api").Span` — is
 * accepted.
 */
export interface OpenTelemetrySpanLike {
  setAttributes(attributes: Record<string, string>): unknown;
}

/**
 * Headers carrying Opik distributed trace context. Accepts both plain
 * dictionaries and any iterable of `[name, value]` pairs (e.g. the WHATWG
 * `Headers` object).
 */
export type HttpHeadersLike =
  | Record<string, string | string[] | undefined>
  | Iterable<[string, string]>;

function getHeader(
  headers: HttpHeadersLike,
  name: string
): string | undefined {
  const target = name.toLowerCase();
  if (
    typeof (headers as Iterable<[string, string]>)[Symbol.iterator] ===
    "function"
  ) {
    for (const [key, value] of headers as Iterable<[string, string]>) {
      if (key.toLowerCase() === target) {
        return value;
      }
    }
    return undefined;
  }
  const record = headers as Record<string, string | string[] | undefined>;
  for (const [key, value] of Object.entries(record)) {
    if (key.toLowerCase() === target) {
      return Array.isArray(value) ? value[0] : value;
    }
  }
  return undefined;
}

/**
 * Extracts Opik distributed trace attributes from HTTP headers.
 *
 * Reads `opik_trace_id` and the optional `opik_parent_span_id` headers
 * (case-insensitive). Values are trimmed; if `opik_trace_id` is missing,
 * empty, whitespace-only, or not a valid UUID the function logs a warning
 * (when invalid, or when `opik_parent_span_id` was provided without a
 * `opik_trace_id`) and returns `null`. If `opik_parent_span_id` is present
 * but not a valid UUID it is dropped with a warning and the trace is still
 * attached using the valid `opik_trace_id`.
 */
export function extractOpikDistributedTraceAttributes(
  httpHeaders: HttpHeadersLike
): OpikDistributedTraceAttributes | null {
  const traceId = getHeader(httpHeaders, OPIK_TRACE_ID_HEADER)?.trim();
  let parentSpanId =
    getHeader(httpHeaders, OPIK_PARENT_SPAN_ID_HEADER)?.trim() || undefined;

  if (!traceId) {
    if (parentSpanId !== undefined) {
      logger.warn(
        `Opik distributed trace header '${OPIK_TRACE_ID_HEADER}' is missing while '${OPIK_PARENT_SPAN_ID_HEADER}' is provided; skipping distributed trace processing.`
      );
    }
    return null;
  }
  if (!isValidUuidV7(traceId)) {
    logger.warn(
      `Opik distributed trace header '${OPIK_TRACE_ID_HEADER}' is not a valid UUIDv7; skipping distributed trace processing.`
    );
    return null;
  }

  if (parentSpanId !== undefined && !isValidUuidV7(parentSpanId)) {
    logger.warn(
      `Opik distributed trace header '${OPIK_PARENT_SPAN_ID_HEADER}' is not a valid UUIDv7; ignoring parent span id.`
    );
    parentSpanId = undefined;
  }

  return new OpikDistributedTraceAttributes(traceId, parentSpanId);
}

/**
 * Attaches an Opik distributed trace parent to the provided OpenTelemetry
 * span by extracting trace context from HTTP headers and setting the
 * corresponding span attributes.
 *
 * @returns `true` if headers were found and attributes were set, `false`
 * otherwise.
 */
export function attachToParent(
  span: OpenTelemetrySpanLike,
  httpHeaders: HttpHeadersLike
): boolean {
  const distributedTraceAttributes =
    extractOpikDistributedTraceAttributes(httpHeaders);
  if (distributedTraceAttributes === null) {
    return false;
  }
  const attributes = distributedTraceAttributes.asAttributes();
  // Mint a stable opik.span_id for this boundary span so descendants picked up by
  // OpikSpanProcessor can chain through it (their opik.parent_span_id will reference
  // this value, and the backend uses opik.span_id verbatim — see OpenTelemetryMapper).
  attributes[OPIK_SPAN_ID] = generateId();
  span.setAttributes(attributes);
  return true;
}